### ISSUE TITLE
"range-val-address" has false-positives when pointers are involved

### ISSUE THREAD

**Issue by zimmski** [mgechev/revive#681]
**Describe the bug**
The "range-val-address" rule flags identifiers and field-selectors to variables defined in a for-range header because the address of these variables stays the same. That is true when the data is in that address but not if the data is accessed through a pointer.

**To Reproduce**

There are test cases in "testdata/range-val-address.go" that show that there are false-positives. I copied them into a single file so i can print their final result:

<details>

```
package main

import (
	"fmt"

	"github.com/kr/pretty"
)

func prettyPrint(object interface{}) {
	fmt.Printf("- %# v\n", pretty.LazyFormatter(object))
}

func rangeValAddress6() {
	type v struct {
		id string
	}
	m := []*string{}

	mySlice := []v{{id: "A"}, {id: "B"}, {id: "C"}}
	for _, value := range mySlice {
		m = append(m, &value.id) // MATCH /suspicious assignment of 'value'. range-loop variables always have the same address/
	}
	fmt.Println("rangeValAddress6")
	prettyPrint(m)
}

func rangeValAddress7() {
	type v struct {
		id string
	}
	m := []*string{}

	for _, value := range []v{{id: "A"}, {id: "B"}, {id: "C"}} {
		m = append(m, &value.id) // MATCH /suspicious assignment of 'value'. range-loop variables always have the same address/
	}
	fmt.Println("rangeValAddress7")
	prettyPrint(m)
}

func rangeValAddress8() {
	type v struct {
		id string
	}
	m := []*string{}

	mySlice := []*v{{id: "A"}, {id: "B"}, {id: "C"}}
	for _, value := range mySlice {
		m = append(m, &value.id)
	}
	fmt.Println("rangeValAddress8")
	prettyPrint(m)
}

func rangeValAddress9() {
	type v struct {
		id string
	}
	m := []*string{}

	mySlice := map[string]*v{"a": {id: "A"}, "b": {id: "B"}, "c": {id: "C"}}
	for _, value := range mySlice {
		m = append(m, &value.id)
	}
	fmt.Println("rangeValAddress9")
	prettyPrint(m)
}

func rangeValAddress10() {
	type v struct {
		id string
	}
	m := []*string{}

	for _, value := range map[string]*v{"a": {id: "A"}, "b": {id: "B"}, "c": {id: "C"}} {
		m = append(m, &value.id)
	}
	fmt.Println("rangeValAddress10")
	prettyPrint(m)
}

func rangeValAddress11() {
	type v struct {
		id string
	}
	m := map[string]*string{}

	for key, value := range map[string]*v{"a": {id: "A"}, "b": {id: "B"}, "c": {id: "C"}} {
		m[key] = &value.id
	}
	fmt.Println("rangeValAddress11")
	prettyPrint(m)
}

func rangeValAddress12() {
	type v struct {
		id string
	}
	m := map[string]*string{}

	for key, value := range map[string]v{"a": {id: "A"}, "b": {id: "B"}, "c": {id: "C"}} {
		m[key] = &value.id // MATCH /suspicious assignment of 'value'. range-loop variables always have the same address/
	}
	fmt.Println("rangeValAddress12")
	prettyPrint(m)
}

func rangeValAddress13() {
	type v struct {
		id string
	}
	m := []*string{}

	otherSlice := map[string]*v{"a": {id: "A"}, "b": {id: "B"}, "c": {id: "C"}}
	mySlice := otherSlice
	for _, value := range mySlice {
		m = append(m, &value.id)
	}
	fmt.Println("rangeValAddress13")
	prettyPrint(m)
}

func main() {
	rangeValAddress6()
	rangeValAddress7()
	rangeValAddress8()
	rangeValAddress9()
	rangeValAddress10()
	rangeValAddress11()
	rangeValAddress12()
	rangeValAddress13()
}
```

</details>

When you run this program you see the following output:

```
rangeValAddress6
- []*string{
    &"C",
    &"C",
    &"C",
}
rangeValAddress7
- []*string{
    &"C",
    &"C",
    &"C",
}
rangeValAddress8
- []*string{
    &"A",
    &"B",
    &"C",
}
rangeValAddress9
- []*string{
    &"A",
    &"B",
    &"C",
}
rangeValAddress10
- []*string{
    &"A",
    &"B",
    &"C",
}
rangeValAddress11
- map[string]*string{
    "a": &"A",
    "b": &"B",
    "c": &"C",
}
rangeValAddress12
- map[string]*string{
    "a": &"C",
    "b": &"C",
    "c": &"C",
}
rangeValAddress13
- []*string{
    &"B",
    &"C",
    &"A",
}
```

So you can see that rangeValAddress8, rangeValAddress9, rangeValAddress10, rangeValAddress11 and rangeValAddress13 are false-positives.

**Comment by zimmski** [comment]
Will do a change for this behavior now, since i cannot merge a PR on our side. If you agree with this report i will do a PR.

**Comment by zimmski** [comment]
Oh wow. Never mind. I just saw that this is all connected to the problem i have with #680. There are no false-positives reported, and they are not marked in those functions. Sorry for the spam.
